File: src/main/host/syscall/futex.c
/*
 * The Shadow Simulator
 * See LICENSE for licensing information
 */

#include "main/host/syscall/futex.h"

#include <errno.h>
#include <inttypes.h>
#include <linux/futex.h>
#include <stdbool.h>
#include <sys/time.h>

#include "lib/logger/logger.h"
#include "main/bindings/c/bindings.h"
#include "main/host/futex.h"
#include "main/host/futex_table.h"
#include "main/host/syscall/protected.h"
#include "main/host/syscall_condition.h"
#include "main/utility/utility.h"

///////////////////////////////////////////////////////////
// Helpers
///////////////////////////////////////////////////////////

static SyscallReturn _syscallhandler_futexWaitHelper(SyscallHandler* sys,
                                                     UntypedForeignPtr futexVPtr, int expectedVal,
                                                     UntypedForeignPtr timeoutVPtr,
                                                     TimeoutType type) {
    // This is a new wait operation on the futex for this thread.
    // Check if a timeout was given in the syscall args.
    CSimulationTime timeoutSimTime = SIMTIME_INVALID;
    if (timeoutVPtr.val) {
        struct timespec ts = {0};
        int rv = process_readPtr(rustsyscallhandler_getProcess(sys), &ts, timeoutVPtr, sizeof(ts));
        if (rv < 0) {
            return syscallreturn_makeDoneErrno(-rv);
        }
        timeoutSimTime = simtime_from_timespec(ts);
        if (timeoutSimTime == SIMTIME_INVALID) {
            return syscallreturn_makeDoneErrno(EINVAL);
        }
    }

    // Normally, the load/compare is done atomically. Since Shadow does not run multiple
    // threads from the same plugin at the same time, we do not use atomic ops.
    // `man 2 futex`: blocking via a futex is an atomic compare-and-block operation
    uint32_t futexVal;
    int result =
        process_readPtr(rustsyscallhandler_getProcess(sys), &futexVal, futexVPtr, sizeof(futexVal));
    if (result) {
        warning("Couldn't read futex address %p", (void*)futexVPtr.val);
        return syscallreturn_makeDoneErrno(-result);
    }

    trace(
        "Futex value is %" PRIu32 ", expected value is %" PRIu32, futexVal, (uint32_t)expectedVal);
    if (!rustsyscallhandler_wasBlocked(sys) && futexVal != (uint32_t)expectedVal) {
        trace("Futex values don't match, try again later");
        return syscallreturn_makeDoneErrno(EAGAIN);
    }

    // Convert the virtual ptr to a physical ptr that can uniquely identify the futex
    ManagedPhysicalMemoryAddr futexPPtr =
        process_getPhysicalAddress(rustsyscallhandler_getProcess(sys), futexVPtr);

    // Check if we already have a futex
    FutexTable* ftable = host_getFutexTable(rustsyscallhandler_getHost(sys));
    Futex* futex = futextable_get(ftable, futexPPtr);

    if (rustsyscallhandler_wasBlocked(sys)) {
        utility_debugAssert(futex != NULL);
        int result = 0;

        // We already blocked on wait, so this is either a timeout or wakeup
        if (timeoutSimTime != SIMTIME_INVALID && rustsyscallhandler_didListenTimeoutExpire(sys)) {
            // Timeout while waiting for a wakeup
            trace("Futex %p timeout out while waiting", (void*)futexPPtr.val);
            result = -ETIMEDOUT;
        } else if (thread_unblockedSignalPending(
                       rustsyscallhandler_getThread(sys),
                       host_getShimShmemLock(rustsyscallhandler_getHost(sys)))) {
            trace("Futex %p has been interrupted by a signal", (void*)futexPPtr.val);
            result = -EINTR;
        } else {
            // Proper wakeup from another thread
            trace("Futex %p has been woke up", (void*)futexPPtr.val);
            result = 0;
        }

        // Dynamically clean up the futex if needed
        if (futex_getListenerCount(futex) == 0) {
            trace("Dynamically freed a futex object for futex addr %p", (void*)futexPPtr.val);
            bool success = futextable_remove(ftable, futex);
            utility_debugAssert(success);
        }

        return syscallreturn_makeDoneI64(result);
    }

    // We'll need to block, dynamically create a futex if one does not yet exist
    if (!futex) {
        trace("Dynamically created a new futex object for futex addr %p", (void*)futexPPtr.val);
        futex = futex_new(futexPPtr);
        bool success = futextable_add(ftable, futex);
        utility_debugAssert(success);
    }

    // Now we need to block until another thread does a wake on the futex.
    trace("Futex blocking for wakeup %s timeout",
          timeoutSimTime != SIMTIME_INVALID ? "with" : "without");
    Trigger trigger =
        (Trigger){.type = TRIGGER_FUTEX, .object = futex, .status = STATUS_FUTEX_WAKEUP};
    SysCallCondition* cond = syscallcondition_new(trigger);
    if (timeoutSimTime != SIMTIME_INVALID) {
        CEmulatedTime timeoutEmulatedTime = (type == TIMEOUT_RELATIVE)
                                                ? timeoutSimTime + worker_getCurrentEmulatedTime()
                                                : timeoutSimTime;
        syscallcondition_setTimeout(cond, timeoutEmulatedTime);
    }
    return syscallreturn_makeBlocked(cond, true);
}

static SyscallReturn _syscallhandler_futexWakeHelper(SyscallHandler* sys,
                                                     UntypedForeignPtr futexVPtr, int numWakeups) {
    // Convert the virtual ptr to a physical ptr that can uniquely identify the futex
    ManagedPhysicalMemoryAddr futexPPtr =
        process_getPhysicalAddress(rustsyscallhandler_getProcess(sys), futexVPtr);

    // Lookup the futex in the futex table
    FutexTable* ftable = host_getFutexTable(rustsyscallhandler_getHost(sys));
    Futex* futex = futextable_get(ftable, futexPPtr);

    trace("Found futex %p at futex addr %p", futex, (void*)futexPPtr.val);

    unsigned int numWoken = 0;
    if (futex && numWakeups > 0) {
        trace("Futex trying to perform %i wakeups", numWakeups);
        numWoken = futex_wake(futex, (unsigned int)numWakeups);
        trace("Futex was able to perform %i/%i wakeups", numWoken, numWakeups);
    }

    return syscallreturn_makeDoneU64(numWoken);
}

///////////////////////////////////////////////////////////
// System Calls
///////////////////////////////////////////////////////////

// TODO: currently only supports uaddr from the same virtual address space (i.e., threads)
// Support across different address spaces requires us to compute a unique id from the
// hardware address (i.e., page table and offset). This is needed, e.g., when using
// futexes across process boundaries.
SyscallReturn syscallhandler_futex(SyscallHandler* sys, const SysCallArgs* args) {
    utility_debugAssert(sys && args);

    UntypedForeignPtr uaddrptr = args->args[0].as_ptr; // int*
    int futex_op = args->args[1].as_i64;
    int val = args->args[2].as_i64;
    UntypedForeignPtr timeoutptr = args->args[3].as_ptr; // const struct timespec*, or uint32_t
    UntypedForeignPtr uaddr2ptr = args->args[4].as_ptr;  // int*
    int val3 = args->args[5].as_i64;

    const int possible_options = FUTEX_PRIVATE_FLAG | FUTEX_CLOCK_REALTIME;
    int options = futex_op & possible_options;
    int operation = futex_op & ~possible_options;

    trace("futex called with addr=%p op=%i (operation=%i and options=%i) and val=%i",
          (void*)uaddrptr.val, futex_op, operation, options, val);

    switch (operation) {
        case FUTEX_WAIT: {
            trace("Handling FUTEX_WAIT operation %i", operation);
            return _syscallhandler_futexWaitHelper(
                sys, uaddrptr, val, timeoutptr, TIMEOUT_RELATIVE);
        }

        case FUTEX_WAKE: {
            trace("Handling FUTEX_WAKE operation %i", operation);
            return _syscallhandler_futexWakeHelper(sys, uaddrptr, val);
        }

        case FUTEX_WAIT_BITSET: {
            trace("Handling FUTEX_WAIT_BITSET operation %i bitset %d", operation, val3);
            if (val3 == FUTEX_BITSET_MATCH_ANY) {
                return _syscallhandler_futexWaitHelper(
                    sys, uaddrptr, val, timeoutptr, TIMEOUT_ABSOLUTE);
            }
            // Other bitsets not yet handled.
            break;
        }
        case FUTEX_WAKE_BITSET: {
            trace("Handling FUTEX_WAKE_BITSET operation %i bitset %d", operation, val3);
            if (val3 == FUTEX_BITSET_MATCH_ANY) {
                return _syscallhandler_futexWakeHelper(sys, uaddrptr, val);
            }
            // Other bitsets not yet handled.
            break;
        }

        case FUTEX_FD:
        case FUTEX_REQUEUE:
        case FUTEX_WAKE_OP:
        case FUTEX_LOCK_PI:
        case FUTEX_TRYLOCK_PI:
        case FUTEX_UNLOCK_PI:
        case FUTEX_CMP_REQUEUE_PI:
        case FUTEX_WAIT_REQUEUE_PI: break;
    }

    warning("Unhandled futex operation %i", operation);
    return syscallreturn_makeDoneErrno(ENOSYS);
}

File: src/main/host/syscall/handler/futex.rs
use linux_api::errno::Errno;
use linux_api::futex::{FUTEX_BITSET_MATCH_ANY, FutexOpFlags};
use shadow_shim_helper_rs::emulated_time::EmulatedTime;
use shadow_shim_helper_rs::simulation_time::SimulationTime;
use shadow_shim_helper_rs::syscall_types::ForeignPtr;

use crate::core::worker::Worker;
use crate::cshadow as c;
use crate::host::futex_table::FutexRef;
use crate::host::syscall::handler::{SyscallContext, SyscallHandler};
use crate::host::syscall::type_formatting::SyscallNonDeterministicArg;
use crate::host::syscall::types::SyscallError;

impl SyscallHandler {
    log_syscall!(
        futex,
        /* rv */ std::ffi::c_int,
        /* uaddr */ *const u32,
        /* op */ std::ffi::c_int,
        /* val */ u32,
        /* utime */ *const std::ffi::c_void,
        /* uaddr2 */ *const u32,
        /* val3 */ SyscallNonDeterministicArg<u32>,
    );
    pub fn futex(
        ctx: &mut SyscallContext,
        uaddr: ForeignPtr<u32>,
        op: std::ffi::c_int,
        val: u32,
        utime: ForeignPtr<linux_api::time::kernel_timespec>,
        _uaddr2: ForeignPtr<u32>,
        val3: u32,
    ) -> Result<std::ffi::c_int, SyscallError> {
        // TODO: currently only supports uaddr from the same virtual address space (i.e., process)
        // Support across different address spaces requires us to compute a unique id from the
        // hardware address (i.e., page table and offset). This is needed, e.g., when using
        // futexes across process boundaries.

        let op = FutexOpFlags::from_bits_retain(op);

        const POSSIBLE_OPTIONS: FutexOpFlags =
            FutexOpFlags::FUTEX_PRIVATE_FLAG.union(FutexOpFlags::FUTEX_CLOCK_REALTIME);
        let options = op.intersection(POSSIBLE_OPTIONS);
        let operation = op.difference(POSSIBLE_OPTIONS);

        log::trace!(
            "futex called with addr={uaddr:p} op={op:?} (operation={operation:?} and options={options:?}) and val={val}",
        );

        match operation {
            FutexOpFlags::FUTEX_WAIT => {
                log::trace!("Handling FUTEX_WAIT operation {operation:?}");
                return Self::futex_wait_helper(ctx, uaddr, val, utime, TimeoutType::Relative);
            }
            FutexOpFlags::FUTEX_WAKE => {
                log::trace!("Handling FUTEX_WAKE operation {operation:?}");
                // TODO: Should we do better than a cast here? Maybe should add a test for this,
                // and/or warn if it overflows?
                return Ok(Self::futex_wake_helper(ctx, uaddr.cast::<()>(), val) as i32);
            }
            FutexOpFlags::FUTEX_WAIT_BITSET => {
                log::trace!("Handling FUTEX_WAIT_BITSET operation {operation:?} bitset {val3:b}");
                if val3 == FUTEX_BITSET_MATCH_ANY {
                    return Self::futex_wait_helper(ctx, uaddr, val, utime, TimeoutType::Absolute);
                }
                // Other bitsets not yet handled.
            }
            FutexOpFlags::FUTEX_WAKE_BITSET => {
                log::trace!("Handling FUTEX_WAKE_BITSET operation {operation:?} bitset {val3:b}");
                if val3 == FUTEX_BITSET_MATCH_ANY {
                    // TODO: Should we do better than a cast here? Maybe should add a test for this,
                    // and/or warn if it overflows?
                    return Ok(Self::futex_wake_helper(ctx, uaddr.cast::<()>(), val) as i32);
                }
                // Other bitsets not yet handled.
            }
            _ => {}
        }

        log::warn!("Unhandled futex operation {operation:?}");
        Err(Errno::ENOSYS.into())
    }

    fn futex_wake_helper(ctx: &mut SyscallContext, ptr: ForeignPtr<()>, num_wakeups: u32) -> u32 {
        // convert the virtual ptr to a physical ptr that can uniquely identify the futex
        let ptr = ctx.objs.process.physical_address(ptr);

        // lookup the futex in the futex table
        let table = ctx.objs.host.futextable_borrow();
        let futex = table.get(ptr);

        let Some(futex) = futex else {
            log::trace!("No futex found at futex addr {ptr:p}");
            return 0;
        };

        log::trace!("Found futex {:p} at futex addr {ptr:p}", futex.ptr());

        if num_wakeups == 0 {
            return 0;
        }

        log::trace!("Futex trying to perform {num_wakeups} wakeups");
        let num_woken = futex.wake(num_wakeups);
        log::trace!("Futex was able to perform {num_woken}/{num_wakeups} wakeups");

        num_woken
    }

    fn futex_wait_helper(
        ctx: &mut SyscallContext,
        ptr: ForeignPtr<u32>,
        expected_val: u32,
        timeout: ForeignPtr<linux_api::time::kernel_timespec>,
        timeout_type: TimeoutType,
    ) -> Result<i32, SyscallError> {
        let mem = ctx.objs.process.memory_borrow();

        // This is a new wait operation on the futex for this thread.
        // Check if a timeout was given in the syscall args.
        let timeout = if timeout.is_null() {
            None
        } else {
            let tspec = mem.read(timeout)?;
            let sim_time = SimulationTime::try_from(tspec).map_err(|_| Errno::EINVAL)?;
            Some(sim_time)
        };

        // Normally, the load/compare is done atomically. Since Shadow does not run multiple
        // threads from the same plugin at the same time, we do not use atomic ops.
        // `man 2 futex`: blocking via a futex is an atomic compare-and-block operation
        let futex_val = mem.read(ptr)?;

        log::trace!("Futex value is {futex_val}, expected value is {expected_val}");
        if !ctx.handler.is_blocked() && futex_val != expected_val {
            log::trace!("Futex values don't match, try again later");
            return Err(Errno::EAGAIN.into());
        }

        // convert the virtual ptr to a physical ptr that can uniquely identify the futex
        let ptr = ctx.objs.process.physical_address(ptr.cast::<()>());

        // check if we already have a futex
        let mut table = ctx.objs.host.futextable_borrow_mut();
        let futex = table.get(ptr);

        if ctx.handler.is_blocked() {
            let futex = futex.expect("syscall was blocked, but there wasn't an existing futex");

            let result;

            // we already blocked on wait, so this is either a timeout or wakeup
            if timeout.is_some() && ctx.handler.did_listen_timeout_expire() {
                // timeout while waiting for a wakeup
                log::trace!("Futex {ptr:p} timeout out while waiting");
                result = Err(Errno::ETIMEDOUT);
            } else if ctx.objs.thread.unblocked_signal_pending(
                ctx.objs.process,
                &ctx.objs.host.shim_shmem_lock_borrow().unwrap(),
            ) {
                log::trace!("Futex {ptr:p} has been interrupted by a signal");
                result = Err(Errno::EINTR);
            } else {
                // proper wakeup from another thread
                log::trace!("Futex {ptr:p} has been woken up");
                result = Ok(0);
            }

            // dynamically clean up the futex if needed
            if futex.listener_count() == 0 {
                log::trace!("Dynamically freed a futex object for futex addr {ptr:p}");
                table.remove(ptr).expect("futex disappeared");
            }

            return result.map_err(Into::into);
        }

        // we'll need to block; dynamically create a futex if one does not yet exist
        let futex = match futex {
            Some(x) => x.clone(),
            None => {
                log::trace!("Dynamically created a new futex object for futex addr {ptr:p}");

                let futex = unsafe { c::futex_new(ptr) };
                assert!(!futex.is_null());
                let futex = unsafe { FutexRef::new(futex) };

                table
                    .add(futex.clone())
                    .expect("new futex is already in table");

                futex
            }
        };

        // now we need to block until another thread does a wake on the futex
        log::trace!(
            "Futex blocking for wakeup {} timeout",
            if timeout.is_some() { "with" } else { "without" },
        );
        let mut rv = SyscallError::new_blocked_on_futex(futex, /* restartable= */ true);
        if let Some(timeout) = timeout {
            let now = Worker::current_time().unwrap();
            let timeout = match timeout_type {
                TimeoutType::Relative => now + timeout,
                TimeoutType::Absolute => EmulatedTime::UNIX_EPOCH + timeout,
            };

            // handle if the timeout has already expired
            let timeout = std::cmp::max(timeout, now);

            rv.blocked_condition().unwrap().set_timeout(Some(timeout));
        }

        Err(rv)
    }

    log_syscall!(
        get_robust_list,
        /* rv */ std::ffi::c_int,
        /* pid */ std::ffi::c_int,
        /* head_ptr */ *const std::ffi::c_void,
        /* len_ptr */ *const libc::size_t,
    );
    pub fn get_robust_list(
        _ctx: &mut SyscallContext,
        _pid: std::ffi::c_int,
        _head_ptr: ForeignPtr<ForeignPtr<linux_api::futex::robust_list_head>>,
        _len_ptr: ForeignPtr<libc::size_t>,
    ) -> Result<(), Errno> {
        warn_once_then_debug!("get_robust_list was called but we don't yet support it");
        Err(Errno::ENOSYS)
    }

    log_syscall!(
        set_robust_list,
        /* rv */ std::ffi::c_int,
        /* head */ *const std::ffi::c_void,
        /* len */ libc::size_t,
    );
    pub fn set_robust_list(
        _ctx: &mut SyscallContext,
        _head: ForeignPtr<linux_api::futex::robust_list_head>,
        _len: libc::size_t,
    ) -> Result<(), Errno> {
        warn_once_then_debug!("set_robust_list was called but we don't yet support it");
        Err(Errno::ENOSYS)
    }
}

/// The type of futex-wait timeout.
enum TimeoutType {
    /// Timeout is relative to current time.
    Relative,
    /// Timeout is absolute.
    Absolute,
}
